cpp/comms/std/src/ucp_helper.h

```diff
@@ -53,165 +63,174 @@ static const int UCP_ANY_RANK = -1;
 /**
  * @brief Asynchronous send callback sets request to completed
  */
-static void send_handle(void *request, ucs_status_t status) {
+static void send_callback(void *request, ucs_status_t status) {
   struct ucx_context *context = (struct ucx_context *)request;
   context->completed = 1;
 }
 
 /**
  * @brief Asynchronous recv callback sets request to completed
  */
-static void recv_handle(void *request, ucs_status_t status,
-                        ucp_tag_recv_info_t *info) {
+static void recv_callback(void *request, ucs_status_t status,
+                          ucp_tag_recv_info_t *info) {
   struct ucx_context *context = (struct ucx_context *)request;
   context->completed = 1;
 }
 
-void load_ucp_handle(struct comms_ucp_handle *ucp_handle) {
-  ucp_handle->ucp_handle =
-    dlopen("libucp.so", RTLD_LAZY | RTLD_NOLOAD | RTLD_NODELETE);
-  if (!ucp_handle->ucp_handle) {
-    ucp_handle->ucp_handle = dlopen("libucp.so", RTLD_LAZY | RTLD_NODELETE);
-    ASSERT(ucp_handle->ucp_handle, "Cannot open UCX library: %s\n", dlerror());
+/**
+ * Helper class for managing `dlopen` state and
+ * interacting with ucp.
+ */
+class comms_ucp_handler {
+ public:
+  comms_ucp_handler() {
+    load_ucp_handle();
+    load_send_func();
+    load_recv_func();
+    load_free_req_func();
+    load_print_info_func();
+    load_worker_progress_func();
   }
-  dlerror();
-}
-
-void close_ucp_handle(struct comms_ucp_handle *handle) {
-  dlclose(handle->ucp_handle);
-}
 
-void assert_dlerror() {
-  char *error = dlerror();
-  ASSERT(error == NULL, "Error loading function symbol: %s\n", error);
-}
+  ~comms_ucp_handler() { dlclose(ucp_handle); }
 
-void load_send_func(struct comms_ucp_handle *ucp_handle) {
-  ucp_handle->send_func = (ucs_status_ptr_t(*)(
-    ucp_ep_h, const void *, size_t, ucp_datatype_t, ucp_tag_t,
-    ucp_send_callback_t))dlsym(ucp_handle->ucp_handle, "ucp_tag_send_nb");
-  assert_dlerror();
-}
+ private:
+  void *ucp_handle;
 
-void load_free_req_func(struct comms_ucp_handle *ucp_handle) {
-  ucp_handle->req_free_func =
-    (void (*)(void *request))dlsym(ucp_handle->ucp_handle, "ucp_request_free");
-  assert_dlerror();
-}
+  ucs_status_ptr_t (*send_func)(ucp_ep_h, const void *, size_t, ucp_datatype_t,
+                                ucp_tag_t, ucp_send_callback_t);
+  ucs_status_ptr_t (*recv_func)(ucp_worker_h, void *, size_t count,
+                                ucp_datatype_t datatype, ucp_tag_t, ucp_tag_t,
+                                ucp_tag_recv_callback_t);
+  void (*print_info_func)(ucp_ep_h, FILE *);
+  void (*req_free_func)(void *);
+  int (*worker_progress_func)(ucp_worker_h);
```

Comment:
minor nitpick: better thing to do for readability would be to `typedef` these functors. They'll also be useful when trying to intialize these pointers inside `load_*` methods.